JAVA-1917: Add ability to set node on statement

GregBestland opened the pull request:
Add ability to setNode on statement, this buypasses the current query plan and creates a new one that contains only the targeted node.

GregBestland committed:
Add setNode to statement

GregBestland committed:
Update changelog

GregBestland committed:
Adding revapi exceptions

GregBestland committed:
Additional cleanup for IT

olim7t commented on changelog/README.md:
```diff
@@ -9,6 +9,7 @@
 - [improvement] JAVA-1925: Rename context getters
 - [improvement] JAVA-1544: Check API compatibility with Revapi
 - [new feature] JAVA-1900: Add support for virtual tables
+- [new feature] JAVA-1917: Add ability to set node on statement
```
Nit: could you add the ticket at the top of the changelog?

I changed this in 4.x because it feels more natural to put more recent tickets first, since versions are ordered that way. It's also how it's done in the Cassandra project.

Maybe we should adopt this convention for the next 3.x versions too for consistency.

olim7t commented on core/src/main/java/com/datastax/oss/driver/api/core/cql/Statement.java:
```diff
@@ -93,6 +94,12 @@
+  T setNode(@Nullable Node node);
+
+  @Nullable
+  Node getNode();
```
This is a pretty generic feature and it might apply to other types of requests than CQL, so I think `Request` would be a better place for this method (`setNode` can stay here, since `Request` only defines its properties, not how they are set in subtypes).
The only other implementation is `PrepareRequest`, it should never be targeted so `getNode` will return null.

Could you also add javadocs? I think [the ones from the 3.x PR](https://github.com/datastax/java-driver/blob/3.x/driver-core/src/main/java/com/datastax/driver/core/Statement.java#L625-L638) can almost be copied as-is (the part about not propagating to prepared statements would be better in the javadocs of `CqlSession#prepare(SimpleStatement)`, which already detail all other fields).

olim7t commented on core/src/main/java/com/datastax/oss/driver/api/core/cql/BatchStatementBuilder.java:
```diff
@@ -146,6 +146,7 @@ public BatchStatement build() {
         serialConsistencyLevel,
-        timeout);
+        timeout,
+        null);
```
The last argument should be `node` (inherited from the parent builder) instead of `null`.

olim7t commented on core/src/main/java/com/datastax/oss/driver/api/core/cql/StatementBuilder.java:
```diff
@@ -199,6 +202,11 @@ public T withTimeout(@Nullable Duration timeout) {
     return self;
   }
 
+  public T withNode(@Nullable Node node) {
```
Could you link to the setter's javadocs like other methods?

olim7t commented on core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandlerBase.java:
```diff
@@ -153,10 +154,16 @@ protected CqlRequestHandlerBase(
+    if (this.statement.getNode() != null) {
+      this.queryPlan = new ConcurrentLinkedQueue<Node>();
+      this.queryPlan.add(this.statement.getNode());
+
```
Since JAVA-1883 there is a dedicated queue implementation for query plans, we can also use it here:
```java
this.queryPlan = new QueryPlan(this.statement.getNode());
```

olim7t commented on core/src/main/java/com/datastax/oss/driver/internal/core/cql/DefaultPreparedStatement.java:
```diff
@@ -103,7 +103,6 @@ public DefaultPreparedStatement(
     this.areBoundStatementsTracing = areBoundStatementsTracing;
-
     this.codecRegistry = codecRegistry;
     this.protocolVersion = protocolVersion;
```
Nit: could you avoid modifying unrelated lines? This adds noise when you do a git blame on the file later (admittedly not for deleted lines, but still).

olim7t commented on integration-tests/src/test/java/com/datastax/oss/driver/api/core/loadbalancing/NodeTargetingTestIT.java:
```diff
@@ -0,0 +1,115 @@
+import org.junit.experimental.categories.Category;
+
+@Category(ParallelizableTests.class)
+public class NodeTargetingTestIT {
```
Nit: no need to repeat "Test" before "IT".

olim7t commented on integration-tests/src/test/java/com/datastax/oss/driver/api/core/loadbalancing/NodeTargetingTestIT.java:
```diff
@@ -0,0 +1,115 @@
+  @Test
+  public void should_use_host_on_statement() {
+    Collection<Node> nodeCol = sessionRule.session().getMetadata().getNodes().values();
+    List<Node> nodes = new ArrayList(nodeCol);
```
`new ArrayList<>(nodeCol)` to avoid the warning (same in the two other methods).

olim7t commented on integration-tests/src/test/java/com/datastax/oss/driver/api/core/loadbalancing/NodeTargetingTestIT.java:
```diff
@@ -0,0 +1,115 @@
+      // when statement is executed
+      ResultSet result = sessionRule.session().execute(statement);
+
+      Node coordinator = result.getExecutionInfo().getCoordinator();
```
unused variable

olim7t commented on integration-tests/src/test/java/com/datastax/oss/driver/api/core/loadbalancing/NodeTargetingTestIT.java:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  @Test
+  public void should_use_host_on_statement() {
```
Nit: 4.x uses the terminology "node" exclusively (I did that to avoid any ambiguity with `InetSocketAddress.getHostName`, with CASSANDRA-7544 there can now be multiple nodes on the same host).

olim7t commented on integration-tests/src/test/java/com/datastax/oss/driver/api/core/loadbalancing/NodeTargetingTestIT.java:
```diff
@@ -0,0 +1,115 @@
+    Statement statement = SimpleStatement.newInstance(query).setNode(node1);
+    // when statement is executed an error should be raised.
+    try {
+      ResultSet result = sessionRule.session().execute(statement);
```
No need to store the result in a variable here.

olim7t commented on integration-tests/src/test/java/com/datastax/oss/driver/api/core/loadbalancing/NodeTargetingTestIT.java:
```diff
@@ -0,0 +1,115 @@
+  public void should_fail_if_host_is_not_connected() {
+    // given a statement with host explicitly set that for which we have no active pool.
+    simulacron.cluster().node(4).close();
+    ;
```
Unnecessary semicolumn here and at line 113.

olim7t commented on integration-tests/src/test/java/com/datastax/oss/driver/api/core/loadbalancing/NodeTargetingTestIT.java:
```diff
@@ -0,0 +1,115 @@
+      // when statement is executed
+      sessionRule.session().execute(statement);
+      fail("Query should have failed");
+    } catch (AllNodesFailedException e) {
```
More precisely, it's a `NoNodeAvailableException` (a subclass).

Which IMO is also more clear for the end user: "no node was available" and you targeted a node => the targeted node was not available.

olim7t reviewed (changes requested):
Looks good, see comments inline.

+ an additional one: I think `BoundStatementBuilder(BoundStatement)` should consider the template statement's node like the other builders.

GregBestland committed:
PR feedback

GregBestland committed:
Small formatting tweak, Revapi update.

GregBestland committed:
Updating IT to avoid timing issues in jenkins.

GregBestland committed:
Timing fix for IT

olim7t committed:
[fixup] copy template's node in BoundStatementBuilder(BoundStatement)

olim7t committed:
[fixup] consistently use node instead of host

olim7t committed:
[fixup] fix error name in assertion message

olim7t committed:
[fixup] add missing error class in javadoc + import linked classes

olim7t committed:
[fixup] remove misleading `@return` javadoc

As an interface, Statement should not make any assumption on the
returned instance. In fact, the default implementation is immutable and
the result is a copy.

olim7t committed:
[fixup] don't copy node to prepare request

olim7t committed:
[fixup] document that getNode is not propagated to bound statements

GregBestland committed:
Moving Node stop to before, to prevent timing issues.

GregBestland committed:
Merge branch '4.x' into java1917
